File: 0_100/042_trapping_rain_water/Solution.java
class Solution {
     public int trap(int[] height) {
        int n = height.length, ans = 0;
        if (n > 0) {
            int left = 0, right = n - 1;
            int l_max = height[0], r_max = height[n - 1];
            while (left <= right) {
                l_max = l_max > height[left] ? l_max : height[left];
                r_max = r_max > height[right] ? r_max : height[right]; 
                if (l_max < r_max) {
                    ans += l_max - height[left++];
                } else {
                    ans += r_max - height[right--];
                }
            }
        }
        return ans;
    }
}

File: 0_100/042_trapping_rain_water/Solution.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let (mut ans, n) = (0_i32, height.len());
        if n > 0 {
            let (mut left, mut right) = (0, n as i32 - 1);
            let (mut l_max, mut r_max) = (height[0], height[n - 1]);
            while left <= right {
                l_max = l_max.max(height[left as usize]);
                r_max = r_max.max(height[right as usize]);
                if l_max < r_max {
                    ans += l_max - height[left as usize];
                    left += 1;
                } else {
                    ans += r_max - height[right as usize];
                    right -= 1;
                }
            }
        }
        ans
    }
}